AX: iframe within inert, hidden element is exposed to ATs

hoffmanjoshua opened the pull request:
#### 3d864212d2b1150cee5712c950dceac2b52b6b83
<pre>
AX: iframe within inert, hidden element is exposed to ATs
<a href="https://bugs.webkit.org/show_bug.cgi?id=278370">https://bugs.webkit.org/show_bug.cgi?id=278370</a>
<a href="https://rdar.apple.com/134318347">rdar://134318347</a>

Reviewed by Tyler Wilcock.

This patch fixes an issue where iFrames with a style of `visibility: hidden` or the inert attribute,
both of which should hide the element and its children to ATs, fails to hide the iFrame&apos;s children.

To fix this, isIgnored() needs to iterate up the web areas and their iFrame renderers, to check whether
they are non-visible or if they are effectively inert. These properties/styles don&apos;t get passed from
the iFrame&apos;s parent document to the iFrame&apos;s document, so this iteration is necessary.

For the isolated tree to update properly, the children of the iframe need to dirty its subtree. However,
when AXObjectCache::handleChildrenChanged tries to handle a ScrollView (which happens when attributes on
the iFrame element are changed), we bail before dirtying the subtree because scroll views don&apos;t have
nodes nor renderers. To resolve this, this patch adds logic in handleChildrenChanged to propogate
setNeedsToUpdateSubtree and setNeedsToUpdateChildren down to the iFrame&apos;s WebArea.

* LayoutTests/accessibility/iframe-content-inert-expected.txt: Added.
* LayoutTests/accessibility/iframe-content-inert.html: Added.
* LayoutTests/accessibility/iframe-content-visibility-expected.txt: Added.
* LayoutTests/accessibility/iframe-content-visibility.html: Added.
* LayoutTests/accessibility/resources/iframe-button.html: Added.

New tests to test `visibility: hidden`, `inert` style/attribute on iFrames.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleChildrenChanged):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::defaultObjectInclusion const):
(WebCore::AccessibilityObject::webArea const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::shouldWebAreaExposeChildren const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::shouldWebAreaExposeChildren const):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::updateScrollbars):
(WebCore::AccessibilityScrollView::removeChildScrollbar):
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::styleDidChange):

Canonical link: <a href="https://commits.webkit.org/283416@main">https://commits.webkit.org/283416@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b2e259b98ec20ea6a2ab68555d924f15665bed4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66068 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45441 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70100 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16678 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68186 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53240 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16959 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53032 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11614 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69135 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41926 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57198 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33664 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38597 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14575 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15554 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60489 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14922 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71802 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10023 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14335 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60351 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10055 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57260 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60642 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14598 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8280 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1919 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41249 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42325 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43508 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42069 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+
+    var body = document.getElementById("body");
+    body.focus();
+    webArea = accessibilityController.focusedElement;
```
You can use `accessibilityController.accessibleElementById` if you give the body `role="group"` rather than using focus.

twilco commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+    var body = document.getElementById("body");
+    body.focus();
+    webArea = accessibilityController.focusedElement;
+    iframeElement1 = webArea.childAtIndex(0);
```
Calling this `iframeElement1` (emphasis on `1`) implies there's more than one iframe element variable, but that doesn't seem to be the case. Also, what actually is this? A web area? Wonder if a name like `axIframeWebArea` is more appropriate ("element" implies DOM element to me, although I realize we call our test objects "AccessibilityUIElement").

hoffmanjoshua replied:
Yeah this is a web area, I can rename it

twilco commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+    iframeButton1 = accessibilityController.accessibleElementById('frame-button-1');
+
+    setTimeout(async function() {
+        // Wait for the second iFrame to load.
```
I think the capitalization is typically iframe and not iFrame.

twilco commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+
+        await waitFor(() => {
+            iframeElement1 = webArea.childAtIndex(0);
+            return !iframeElement1 || iframeElement1.isIgnored;
```
Do we need to check `isIgnored`? Shouldn't it just always be null if we behave correctly? (Because we remove it from `webArea`'s children)

twilco commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+            return !iframeElement1 || iframeElement1.isIgnored;
+        });
+
+        output += expect("!iframeElement1 || iframeElement1.isIgnored", "true");
```
This can all be `output += await expectAsync("!webArea.childAtIndex(0)", "true");`, though we should add a comment saying something to the effect of "The iframe webarea (if that's what this is) should be removed from the accessibility tree after inert was added"

twilco commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+             iframeButton1 = accessibilityController.accessibleElementById('frame-button-1');
+             return !iframeButton1 || iframeButton1.isIgnored;
+        });
+        output += expect("!iframeButton1 || iframeButton1.isIgnored", "true");
```
Can this be:

`output += await expectAsync("!accessibilityController.accessibleElementById('frame-button-1')", "true");`

twilco commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+             iframeButton2 = accessibilityController.accessibleElementById('frame-button-2');
+             return !iframeButton2 || iframeButton2.isIgnored;
+        });
+        output += expect("!iframeButton2 || iframeButton2.isIgnored", "true");
```
Can this be:

`output += await expectAsync("!accessibilityController.accessibleElementById('frame-button-2')", "true");`

twilco commented on LayoutTests/accessibility/iframe-content-visibility.html:
```diff
@@ -0,0 +1,69 @@
+var output = "This test ensures that iFrame content is ignored when the iFrame has its visibility set to hidden.\n\n";
+var iframeElement1, iframeButton1, iframeButton2, webArea;
+
+function runTest() {
```
This test seems similar to the other, so all review comments should apply to this one too.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1304,6 +1304,11 @@ void AXObjectCache::handleChildrenChanged(AccessibilityObject& object)
+        if (auto* webarea = scrollView->webAreaObject()) {
+            webarea->setNeedsToUpdateSubtree();
+            webarea->setNeedsToUpdateChildren();
+        }
```
Can we narrow this to only being applicable to iframe scrollviews / web areas?

I think we've standardized around capitalizing webarea like webArea.

Do we need to set both `setNeedsToUpdateSubtree` and `setNeedsToUpdateChildren` to fix the bug? Can we add a comment explaining we need to propagate this flag / flags down when iframes change attributes that affect the accessibility tree?

AndresGonzalezApple replied:
auto* webarea --> RefPtr since we are using this pointer after some non-trivial operation.

hoffmanjoshua replied:
> Do we need to set both setNeedsToUpdateSubtree and setNeedsToUpdateChildren to fix the bug? Can we add a comment explaining we need to propagate this flag / flags down when iframes change attributes that affect the accessibility tree?

Just verified that we do. This mimics the behavior of `handleChildrenChanged` beneath the bail out condition that ScrollViews hit—both setNeedsToUpdateSubtree and setNeedsToUpdateChildren are called on `object`

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3995,6 +3995,10 @@ AccessibilityObjectInclusion AccessibilityObject::defaultObjectInclusion() const
 
+    auto* localWebArea = webArea();
+    if (localWebArea && !localWebArea->shouldWebAreaExposeChildren())
+        return AccessibilityObjectInclusion::IgnoreObject;
```
Why "local"? Another option would be `RefPtr webArea = this->webArea();` unless local is really meaningful here.

hoffmanjoshua replied:
Mentioned this elsewhere, but the I wanted a way to semantically differentiate the root web area and this elements parent web area, but using `this->webArea` is probably enough to do that, so I'll update that here.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -4413,6 +4417,22 @@ bool AccessibilityObject::ignoredByRowAncestor() const
+{
+    auto* frameView = documentFrameView();
+    if (!frameView)
+        return nullptr;
```
We can remove this check. `getOrCreate` already does a nullptr check.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -4413,6 +4417,22 @@ bool AccessibilityObject::ignoredByRowAncestor() const
+
+    auto* cache = axObjectCache();
+    if (!cache)
+        return nullptr;
```
We can remove this check.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -4413,6 +4417,22 @@ bool AccessibilityObject::ignoredByRowAncestor() const
+    if (!cache)
+        return nullptr;
+
+    auto* root = cache->getOrCreate(frameView);
```
RefPtr root = cache ? cache->getOrCreate(frameView) : nullptr;

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -4413,6 +4417,22 @@ bool AccessibilityObject::ignoredByRowAncestor() const
+    if (!frameView)
+        return nullptr;
+
+    auto* cache = axObjectCache();
```
We probably want some smart pointer here (e.g. CheckedPtr)

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -4413,6 +4417,22 @@ bool AccessibilityObject::ignoredByRowAncestor() const
 
+AccessibilityObject* AccessibilityObject::webArea() const
+{
+    auto* frameView = documentFrameView();
```
We probably want some smart pointer here (e.g. CheckedPtr)

twilco commented on Source/WebCore/rendering/RenderWidget.cpp:
```diff
@@ -228,8 +228,10 @@ void RenderWidget::styleDidChange(StyleDifference diff, const RenderStyle* oldSt
-        if (CheckedPtr cache = document().existingAXObjectCache())
+        if (CheckedPtr cache = document().existingAXObjectCache()) {
             cache->onWidgetVisibilityChanged(*this);
+            cache->childrenChanged(this);
```
Does this method (`RenderWidget::styleDidChange`) get called when any style on the iframe changes, even ones that don't affect the accessibility tree at all (which would be most of them)? If so, that would mean we're going to issuing potentially a lot more children changed events, most unnecessary.

hoffmanjoshua replied:
This was only needed for the inert attribute, because we weren't handling that in `RenderElement::styleWillChange`. I'm updating `RenderElement::styleWillChange` now to include a check for `inert` changing, like we do for visibility, so we can remove childrenChanged here and only call it once there (and only for those two attributes changing.

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.h:
```diff
@@ -44,6 +44,7 @@ class AccessibilityScrollView final : public AccessibilityObject {
     virtual ~AccessibilityScrollView();
 
     AccessibilityObject* webAreaObject() const override;
+    void setNeedsToUpdateChildren() override { m_childrenDirty = true; }
```
Does this need to be public?

hoffmanjoshua replied:
Yes, now that we are calling it from AXObjectCache we need to make it public here

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -168,6 +178,12 @@ void AccessibilityScrollView::removeChildScrollbar(AccessibilityObject* scrollba
+        if (!cache)
+            return;
+
+        cache->remove(scrollbar->objectID());
```
This can be:

```
if (CheckedPtr cache = axObjectCache())
     cache->remove(scrollbar->objectID());
```

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -147,6 +147,16 @@ void AccessibilityScrollView::updateScrollbars()
     if (!scrollView)
         return;
 
+    auto* localWebArea = webArea();
```
Same question about "local" in the name.

hoffmanjoshua replied:
I am using local here to make it clear that we are referring to the object's parent web area, rather than the root one (since is many places we use the root). But, since we are in a scroll area here, we can probably get away without that.

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.h:
```diff
@@ -138,6 +138,7 @@ class AccessibilityRenderObject : public AccessibilityNodeObject {
     String secureFieldValue() const override;
     void labelText(Vector<AccessibilityText>&) const override;
 
+    bool shouldWebAreaExposeChildren() const override;
```
This method feels a bit awkward to me, it's almost worded like a general utility method, but it's associated with a class, and only does something for `isWebArea()` objects 🤔 

Would it work to instead add a class method to `AccessibilityObject` that is something like "isWithinHiddenWebArea" that uses your new `webArea()` method, and the rest of the logic currently inside `shouldWebAreaExposeChildren`?

hoffmanjoshua replied:
Yeah I like that idea/naming. Will explore this!

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1078,6 +1078,26 @@ static bool webAreaIsPresentational(RenderObject* renderer)
+
+    CheckedPtr renderView = dynamicDowncast<RenderView>(m_renderer.get());
+    CheckedPtr frameRenderer = renderView ? renderView->frameView().frame().ownerRenderer() : nullptr;
+
```
Take it or leave it: This newline feels unnecessary to me, but maybe y'all have other preferences

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1078,6 +1078,26 @@ static bool webAreaIsPresentational(RenderObject* renderer)
+        renderView = frameRenderer->document().renderView();
+        frameRenderer = renderView ? renderView->frameView().frame().ownerRenderer() : nullptr;
+    }
+
```
Take it or leave it: This newline feels unnecessary to me, but maybe y'all have other preferences

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1078,6 +1078,26 @@ static bool webAreaIsPresentational(RenderObject* renderer)
+    CheckedPtr frameRenderer = renderView ? renderView->frameView().frame().ownerRenderer() : nullptr;
+
+    while (frameRenderer) {
+        const auto usedVisibility = frameRenderer->style().usedVisibility();
```
Do we need this local `usedVisibility`?

hoffmanjoshua replied:
I'll move it inline

AndresGonzalezApple commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+    body.focus();
+    webArea = accessibilityController.focusedElement;
+    iframeElement1 = webArea.childAtIndex(0);
+    iframeButton1 = accessibilityController.accessibleElementById('frame-button-1');
```
We use "" first, and '' for nested strings.

AndresGonzalezApple commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+    setTimeout(async function() {
+        // Wait for the second iFrame to load.
+        await waitFor(() => {
+            iframeButton2 = accessibilityController.accessibleElementById('frame-button-2');
```
Dito: '' -> ""

AndresGonzalezApple commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+        output += expect("iframeButton1.isIgnored", "false");
+        output += expect("iframeButton2.isIgnored", "false");
+
+        output += `\n${evalAndReturn("document.getElementById('iframe1').setAttribute('inert', '');")}`;
```
Dito: '' -> ""

AndresGonzalezApple commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+        output += expect("!iframeElement1 || iframeElement1.isIgnored", "true");
+
+        await waitFor(() => { 
+             iframeButton1 = accessibilityController.accessibleElementById('frame-button-1');
```
Dito: '' -> ""

AndresGonzalezApple commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+
+        await waitFor(() => { 
+             iframeButton1 = accessibilityController.accessibleElementById('frame-button-1');
+             return !iframeButton1 || iframeButton1.isIgnored;
```
isIgnored call is superfluous as Tyler pointed out above.

AndresGonzalezApple commented on LayoutTests/accessibility/iframe-content-inert.html:
```diff
@@ -0,0 +1,69 @@
+        output += expect("!iframeButton1 || iframeButton1.isIgnored", "true");
+
+        await waitFor(() => { 
+             iframeButton2 = accessibilityController.accessibleElementById('frame-button-2');
```
'' -> ""

AndresGonzalezApple commented on Source/WebCore/accessibility/AccessibilityObject.h:
```diff
@@ -745,6 +745,8 @@ class AccessibilityObject : public AXCoreObject, public CanMakeWeakPtr<Accessibi
 
     const AccessibilityScrollView* ancestorAccessibilityScrollView(bool includeSelf) const;
     virtual AccessibilityObject* webAreaObject() const { return nullptr; }
+    virtual bool shouldWebAreaExposeChildren() const { return true; }
```
Is this a webarea specific property? I.e., can it apply to other types of objects?

hoffmanjoshua replied:
No, this is for web areas only since the children will ask for this on their web area. However, I am going to rename this per Tyler's suggestion above.

AndresGonzalezApple commented on Source/WebCore/accessibility/AccessibilityObject.h:
```diff
@@ -745,6 +745,8 @@ class AccessibilityObject : public AXCoreObject, public CanMakeWeakPtr<Accessibi
     virtual AccessibilityObject* webAreaObject() const { return nullptr; }
+    virtual bool shouldWebAreaExposeChildren() const { return true; }
+    AccessibilityObject* webArea() const;
 
```
Now we have webAreObject() and webArea(). Do we need a new method? What's the difference?

hoffmanjoshua replied:
`webAreaObject` only is implemented on AccessibilityScrollView (since that is the only thing that has an associated web area object. We could either:
a) start implementing webareaobject on other AXObjects (though, I wonder if this could get more confusing)
b) rename `webArea` to something like `documentWebArea()` to indicate that this is for any element's web area, not just the scroll views (I'd prefer this)

twilco replied:
In my mind, I read your new `webArea()` as `containingWebArea()`, as in it's the logic for each accessibility object to find its containing web area AX object. This also sufficiently differentiates it from `webAreaObject`, as a scroll view contains a webarea and not the other way around. Further, I don't even think `webAreaObject` needs to be virtual on `AccessibilityObject.h` — any usage really should just be casting to `AccessibilityScrollView` and calling a non-virtual method (which maybe could be better named as `containedWebArea()`, further disambiguating these methods?)

hoffmanjoshua replied:
I like `containedWebArea` a lot, I'll change this to that.

AndresGonzalezApple commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1078,6 +1078,26 @@ static bool webAreaIsPresentational(RenderObject* renderer)
+    }
+
+    return true;
+}
```
We are already doing this type of check in

AccessibilityObjectInclusion AccessibilityObject::defaultObjectInclusion() const
{
if (auto* style = this->style()) {
if (style->effectiveInert())
return AccessibilityObjectInclusion::IgnoreObject;
if (style->usedVisibility() != Visibility::Visible) {
// aria-hidden is meant to override visibility as the determinant in AX hierarchy inclusion.
if (equalLettersIgnoringASCIICase(getAttribute(aria_hiddenAttr), "false"_s))
return AccessibilityObjectInclusion::DefaultBehavior;

return AccessibilityObjectInclusion::IgnoreObject;
}
}

Do we need to do it here again?

hoffmanjoshua replied:
Inside iframes, `effectiveInert()` and `usedVisibility()` won't return the properties if the visibility/inert is set in a parent outside the iframe. That's why we need to do the walkup of the documents/renderers here.

AndresGonzalezApple commented on Source/WebCore/rendering/RenderWidget.cpp:
```diff
@@ -228,8 +228,10 @@ void RenderWidget::styleDidChange(StyleDifference diff, const RenderStyle* oldSt
+        }
     }
 }
 
```
In addition, we should not call cache->childrenChanged and onWidgetVisibilityChanged. We should have only one call to an AXObjectCache event listener, and sort out in there what needs to be updated. Maybe we need to pass more context info to this onWidgetChanged listener to determine what to do.

hoffmanjoshua replied:
I was able to re-write this to remove `childrenChanged` in render widget. We actually could handle that childrenChanged already in `RenderElement`, we just weren't looking for changes to the `inert` attribute yet, so I added in a check for that.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -439,7 +439,7 @@ bool AXObjectCache::isNodeVisible(Node* node) const
     // Check whether this object or any of its ancestors has opacity 0.
     // The resulting opacity of a RenderObject is computed as the multiplication
-    // of its opacity times the opacities of its ancestors.
+    // of its opxacity times the opacities of its ancestors.
```
typo

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1304,6 +1304,13 @@ void AXObjectCache::handleChildrenChanged(AccessibilityObject& object)
         if (auto* parentTable = axRow->parentTable())
             deferRecomputeTableCellSlots(*parentTable);
+    } else if (auto* scrollView = dynamicDowncast<AccessibilityScrollView>(object)) {
+        // When we get a childrenChanged for a iframe (a ScrollView that is not the root web area), then we need to
```
typo: "an iframe"

I feel like "a childrenChanged" is a bit confusing of a noun. Maybe we could say something like "When the children of an iframe change, e.g. because it became visible or non-visible, ...

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1304,6 +1304,13 @@ void AXObjectCache::handleChildrenChanged(AccessibilityObject& object)
             deferRecomputeTableCellSlots(*parentTable);
+    } else if (auto* scrollView = dynamicDowncast<AccessibilityScrollView>(object)) {
+        // When we get a childrenChanged for a iframe (a ScrollView that is not the root web area), then we need to
+        // dirty the web area's subtree since the scroll area doesn't have a node nor renderer
```
It could be nice to explain why not having a node or renderer matters, e.g. "doesn't have a node or renderer, thus failing the check below and returning early."

hoffmanjoshua replied:
I like this, will add to this comment

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3995,6 +3995,9 @@ AccessibilityObjectInclusion AccessibilityObject::defaultObjectInclusion() const
         return AccessibilityObjectInclusion::IgnoreObject;
 
+    if (isWithinHiddenWebArea())
+        return AccessibilityObjectInclusion::IgnoreObject;
```
Maybe combine this with the statement above?

```
if (isARIAHidden() || isWithinHiddenWebArea())
    return AccessibilityObjectInclusion::IgnoreObject;
```

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1304,6 +1304,14 @@ void AXObjectCache::handleChildrenChanged(AccessibilityObject& object)
+        // When the children of an iframe change, e.g., because it's visibility changes,
+        // then we need to dirty the web area's subtree since the scroll area doesn't
+        // have a node nor renderer, thus, failing the check below and returning early.
+        if (RefPtr webArea = scrollView->webAreaObject(); webArea != rootWebArea()) {
```
This is missing a nullptr check. It's one downside of these `;` checks in if-statements — you don't get the implicit null check. `webArea && webArea != rootWebArea()`.

hoffmanjoshua replied:
ah yes I can add that explanation!

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1304,6 +1304,14 @@ void AXObjectCache::handleChildrenChanged(AccessibilityObject& object)
         if (auto* parentTable = axRow->parentTable())
             deferRecomputeTableCellSlots(*parentTable);
+    } else if (auto* scrollView = dynamicDowncast<AccessibilityScrollView>(object)) {
+        // When the children of an iframe change, e.g., because it's visibility changes,
```
Is it it's or its? I think this may have been my typo, sorry 😅

hoffmanjoshua replied:
Will fix!

hoffmanjoshua committed:
AX: iframe within inert, hidden element is exposed to ATs
https://bugs.webkit.org/show_bug.cgi?id=278370
rdar://134318347

Reviewed by Tyler Wilcock.

This patch fixes an issue where iFrames with a style of `visibility: hidden` or the inert attribute,
both of which should hide the element and its children to ATs, fails to hide the iFrame's children.

To fix this, isIgnored() needs to iterate up the web areas and their iFrame renderers, to check whether
they are non-visible or if they are effectively inert. These properties/styles don't get passed from
the iFrame's parent document to the iFrame's document, so this iteration is necessary.

For the isolated tree to update properly, the children of the iframe need to dirty its subtree. However,
when AXObjectCache::handleChildrenChanged tries to handle a ScrollView (which happens when attributes on
the iFrame element are changed), we bail before dirtying the subtree because scroll views don't have
nodes nor renderers. To resolve this, this patch adds logic in handleChildrenChanged to propogate
setNeedsToUpdateSubtree and setNeedsToUpdateChildren down to the iFrame's WebArea.

* LayoutTests/accessibility/iframe-content-inert-expected.txt: Added.
* LayoutTests/accessibility/iframe-content-inert.html: Added.
* LayoutTests/accessibility/iframe-content-visibility-expected.txt: Added.
* LayoutTests/accessibility/iframe-content-visibility.html: Added.
* LayoutTests/accessibility/resources/iframe-button.html: Added.

New tests to test `visibility: hidden`, `inert` style/attribute on iFrames.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleChildrenChanged):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::defaultObjectInclusion const):
(WebCore::AccessibilityObject::webArea const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::shouldWebAreaExposeChildren const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::shouldWebAreaExposeChildren const):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::updateScrollbars):
(WebCore::AccessibilityScrollView::removeChildScrollbar):
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::styleDidChange):

Canonical link: https://commits.webkit.org/283416@main
